### PR TITLE
octopus: os/bluestore: cap omap naming scheme upgrade transactoin.

### PR DESCRIPTION
**Pull request by ifed01**
We shouldn't use single per-onode transaction for such an upgrade when onode's omap list is huge. This results in similarly sized WAL/SST files which are inefficient, might cause high memory usage and sometimes error-prone.

backport tracker: https://tracker.ceph.com/issues/52432
backport of #39377
parent tracker: https://tracker.ceph.com/issues/49170

Signed-off-by: Igor Fedotov <ifedotov@suse.com>
(cherry picked from commit e897fa243c1dd38329733b452872616023f14ac8)

 Conflicts (caused by lack of per-pg omap scheme):
	src/os/bluestore/BlueStore.cc
	src/os/bluestore/BlueStore.h
	src/os/bluestore/bluestore_types.h

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -8173,12 +8182,70 @@ void BlueStore::_fsck_check_object_omap(FSCKDepth depth,
     !o->onode.is_perpool_omap() &&
     !o->onode.is_pgmeta_omap()) {
     dout(10) << "fsck converting " << o->oid << " omap to per-pool" << dendl;
-    bufferlist h;
+    bufferlist header;
     map<string, bufferlist> kv;
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:8186]: kv is unused?

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -3545,6 +3546,57 @@ BlueStore::BlobRef BlueStore::ExtentMap::split_blob(
...
+  std::string* out)
+{
+  if (bluestore_onode_t::is_perpool_omap(flags) &&
+     !bluestore_onode_t::is_pgmeta_omap(flags)) {
+    _key_encode_u64(o->c->pool(), out);
+  }
+  _key_encode_u64(o->onode.nid, out);
+  out->push_back('~');
+}
 //
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:3600]: Should move this code block down, so diff makes more sense.

**Review by aclamk — APPROVED** [review]
Good!
Minor maintenance comments only.

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: yep

**Comment by ideepika** [comment]
`2021-09-09T18:56:25.106 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.101+0000 7f41e955fd80 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck warning: store not yet converted to per-pool stats
2021-09-09T18:56:25.106 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.101+0000 7f41e955fd80 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck warning: store not yet converted to per-pool omap
2021-09-09T18:56:25.126 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.121+0000 7f41d7fb4700 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck warning: #-1:c0371625:::snapmapper:0# has omap that is not per-pool or pgmeta
2021-09-09T18:56:25.279 INFO:teuthology.orchestra.run.smithi204.stdout:Selecting previously unselected package python3-pytest.
2021-09-09T18:56:25.293 INFO:teuthology.orchestra.run.smithi204.stdout:Preparing to unpack .../85-python3-pytest_3.3.2-2_all.deb ...
2021-09-09T18:56:25.338 INFO:teuthology.orchestra.run.smithi204.stdout:Unpacking python3-pytest (3.3.2-2) ...
2021-09-09T18:56:25.578 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.573+0000 7f41d87b5700 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck warning: #2:09401a81:::601.00000000:head# has omap that is not per-pool or pgmeta
2021-09-09T18:56:25.578 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.573+0000 7f41d87b5700 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck warning: #2:d0630e4c:::mds0_sessionmap:head# has omap that is not per-pool or pgmeta
2021-09-09T18:56:25.581 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.577+0000 7f41e955fd80 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck error: legacy statfs record found, removing
2021-09-09T18:56:25.582 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.577+0000 7f41e955fd80 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck error: missing Pool StatFS record for pool 1
2021-09-09T18:56:25.583 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.577+0000 7f41e955fd80 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck error: missing Pool StatFS record for pool 2
2021-09-09T18:56:25.583 INFO:tasks.ceph.osd.11.smithi166.stderr:2021-09-09T18:56:25.577+0000 7f41e955fd80 -1 bluestore(/var/lib/ceph/osd/ceph-11) fsck error: missing Pool StatFS record for pool ffffffffffffffff`

@ifed01 http://qa-proxy.ceph.com/teuthology/yuriw-2021-09-09_16:32:29-rados-wip-yuri7-testing-2021-09-08-1216-octopus-distro-basic-smithi/6381475/teuthology.log seems related to this, can you confirm if it's related or not? thanks!

**Comment by ifed01** [comment]
@neha-ojha @yuriw could you please review QA run I did for this PR? Is this sufficient and OK for proceeding with the merge?
QA runs against this patch:
http://pulpito.front.sepia.ceph.com/?branch=wip-ifed-testing-octopus

The same test cases run against pure Octopus:
http://pulpito.front.sepia.ceph.com/ifed01-2021-11-14_22:05:11-rados-wip-ifed-testing-octopus-pure-distro-default-smithi/
Which indicates the same couple of failing cases.

**Comment by neha-ojha** [comment]
> @neha-ojha @yuriw could you please review QA run I did for this PR? Is this sufficient and OK for proceeding with the merge? QA runs against this patch: http://pulpito.front.sepia.ceph.com/?branch=wip-ifed-testing-octopus
> 
> The same test cases run against pure Octopus: http://pulpito.front.sepia.ceph.com/ifed01-2021-11-14_22:05:11-rados-wip-ifed-testing-octopus-pure-distro-default-smithi/ Which indicates the same couple of failing cases.

@ifed01  In http://pulpito.front.sepia.ceph.com/ifed01-2021-11-15_09:10:19-rados-wip-ifed-testing-octopus-distro-default-smithi/ 

- 6503857 is waiting on a backport https://tracker.ceph.com/issues/52077
- 6503864 has only showed up in your testing https://sentry.ceph.com/organizations/ceph/issues/18984/events/fad8649d68f34a15a4bd5dcbd4586da9/events/?project=2 but does not look related

But, how were these 32 jobs chosen? we usually do broad rados runs for backports, which include ~350 jobs

**Comment by ifed01** [comment]
> 
> @ifed01 In http://pulpito.front.sepia.ceph.com/ifed01-2021-11-15_09:10:19-rados-wip-ifed-testing-octopus-distro-default-smithi/
> 
> * 6503857 is waiting on a backport https://tracker.ceph.com/issues/52077
> * 6503864 has only showed up in your testing https://sentry.ceph.com/organizations/ceph/issues/18984/events/fad8649d68f34a15a4bd5dcbd4586da9/events/?project=2 but does not look related

OK, thanks!

> 
> But, how were these 32 jobs chosen? we usually do broad rados runs for backports, which include ~350 jobs
Just arbitrary limited to this number of jobs for the first attempt. What would be the proper command line to run this sort of tests? 

Would something like the following be OK?
bin/teuthology-suite -s rados -c wip-ifed-testing-octopus -m smithi --subset 1/10000 --ceph-repo git://github.com/ceph/ceph-ci.git --filter-out dashboard,cephadm  -p 200

**Comment by neha-ojha** [comment]
> > @ifed01 In http://pulpito.front.sepia.ceph.com/ifed01-2021-11-15_09:10:19-rados-wip-ifed-testing-octopus-distro-default-smithi/
> > 
> > * 6503857 is waiting on a backport https://tracker.ceph.com/issues/52077
> > * 6503864 has only showed up in your testing https://sentry.ceph.com/organizations/ceph/issues/18984/events/fad8649d68f34a15a4bd5dcbd4586da9/events/?project=2 but does not look related
> 
> OK, thanks!
> 
> > But, how were these 32 jobs chosen? we usually do broad rados runs for backports, which include ~350 jobs
> > Just arbitrary limited to this number of jobs for the first attempt. What would be the proper command line to run this sort of tests?
> 
> Would something like the following be OK? bin/teuthology-suite -s rados -c wip-ifed-testing-octopus -m smithi --subset 1/10000 --ceph-repo git://github.com/ceph/ceph-ci.git --filter-out dashboard,cephadm -p 200

looks good, barring the fact that it may not be a good idea to filter-out cephadm and dashboard, unless you are super sure to not break those suites :)

**Comment by ifed01** [comment]
https://pulpito.ceph.com/nojha-2021-11-23_23:38:20-rados-wip-ifed-testing-octopus-distro-basic-smithi/
https://pulpito.ceph.com/nojha-2021-11-24_17:01:25-rados-wip-ifed-testing-octopus-distro-basic-smithi/?branch=wip-ifed-testing-octopus

**Comment by ifed01** [comment]
jenkins test api

**Pull request merged**